File: LC/Medium/621.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        ans = 0
        count_tasks = Counter(tasks)
        heap = []

        for count in count_tasks.values():
            heapq.heappush(heap, -count)

        elements = []

        while elements or heap:
            ans += 1
            if heap:
                count = heapq.heappop(heap) + 1
                if count < 0:
                    elements.append((count, ans + n))

            if elements and elements[0][1] == ans:
                heapq.heappush(heap, elements.pop(0)[0])

        return ans

File: LC/C++/Medium/621.cpp
#include <vector>
#include <queue>
#include <unordered_map>

class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char, int> count;
        for (char c : tasks) {
            count[c]++;
        }

        priority_queue<int> heap;
        for (auto& p : count) {
            heap.push(p.second);
        }

        queue<pair<int, int>> cooldown;
        int ans = 0;

        while (!heap.empty() || !cooldown.empty()) {
            ++ans;

            if (!heap.empty()) {
                int c = heap.top();
                heap.pop();
                c--;

                if (c > 0) {
                    cooldown.push({c, ans + n});
                }
            }

            if (!cooldown.empty() && cooldown.front().second == ans) {
                heap.push(cooldown.front().first);
                cooldown.pop();
            }
        }

        return ans;
        
    }
};